aten/src/ATen/native/BatchLinearAlgebraKernel.cpp

```diff
@@ -915,12 +914,29 @@ void apply_lu_factor(const Tensor& input, const Tensor& pivots, const Tensor& in
   auto n = input.size(-1);
   auto leading_dimension = std::max<int64_t>(1, m);
 
-  for (const auto i : c10::irange(batch_size)) {
-    scalar_t* input_working_ptr = &input_data[i * input_matrix_stride];
-    int* pivots_working_ptr = &pivots_data[i * pivots_stride];
-    int* infos_working_ptr = &infos_data[i];
-    lapackLu<scalar_t>(m, n, input_working_ptr, leading_dimension, pivots_working_ptr, infos_working_ptr);
-  }
+  const auto loop = [&](int64_t start, int64_t end) {
+    for (const auto i : c10::irange(start, end)) {
+      scalar_t* input_working_ptr = &input_data[i * input_matrix_stride];
+      int* pivots_working_ptr = &pivots_data[i * pivots_stride];
+      int* infos_working_ptr = &infos_data[i];
+      lapackLu<scalar_t>(
+          m,
+          n,
+          input_working_ptr,
+          leading_dimension,
+          pivots_working_ptr,
+          infos_working_ptr);
+    }
+  };
+  int64_t matrix_rank = std::min(m, n);
+  // A threshold test out on 32 core/socket ICX system
+  int64_t grain_size;
+  float log_grain_size = std::log(3200) - 3 * std::log(matrix_rank);
+  if (log_grain_size < 0)
+    grain_size = 1;
+  else
+    grain_size = int64_t(std::exp(log_grain_size));
```

Comment:
Taking logs and then exps is not only slow but it may also lose precision.
```suggestion
  auto matrix_rank = std::min(m, n);
  // A heuristic tested on a 32 core/socket ICX system
  // https://github.com/pytorch/pytorch/pull/93037#discussion_r1090112948
  auto grain_size = int64_t{3200.0 / (matrix_rank * matrix_rank * matrix_rank)}
```

also, more generally, please use braces when using if-else clauses.

Reply:
We used log to avoid overflows, and now the result is being exped to again introduce potential overflow issues...

Reply:
I read it like this: the larger the matrix, the more weight it is to parallelize over batch dimensions. Is this really what we want?

Reply:
It should be, as per https://github.com/pytorch/pytorch/pull/93037#issuecomment-1405007779, right?

Reply:
As I have mentioned above, these benchmarks are just for matrices up to 64, I would prefer to see a more comprehensive test case for different dtypes and sizes.

Reply:
Hi, @nikitaved. I updated the benchmark for bs=32, thread_num=32 (since we set the grain size >=1). And iter the rank from 1024 to 16384.

| rank | 1k |  2k |  4k | 8k |  16k| 
| ---- | --- | --- | --- | --- | --- |
| parallel inside lapack | 1.0476 |2.7907| 11.163|52.867 |358.62 |
| parallel outside lapack | 0.4165 | 1.6828 | 9.1045| 52.794| 347.25 |
|speed up ratio| 2.52x | 1.61x  | 1.22xx|1.001x |1.03x |

It is true "the larger the matrix, the more weight it is for parallelizing over batch dimensions inside lapack". But since we see parallel within `at::parallel` will not worse than parallel inside lapack. So we may not need to set an upper bound to back to parallel inside lapack.
 



```python
# lu_factor.py
import torch
import time

def bench(bs, r):
    x = torch.randn(int(bs), r, r)
    start = time.time()
    for i in range(10):
        y1 = torch.linalg.lu_factor(x)
    end = time.time()
    print(end - start)

for r in [10, 11, 12, 13, 14]:
    rr = 2**r
    for bs in [32]:
        bench(bs, rr)
```
```bash
export LD_PRELOAD=${CONDA_PREFIX}/lib/libjemalloc.so
echo "parallel inside"
OMP_NUM_THREADS=32 numactl -C 0-31 -m 0 python -u lu_factor.py
echo "parallel outside"
LU_TORCH_PARALLEL=1 OMP_NUM_THREADS=32 numactl -C 0-31 -m 0 python -u lu_factor.py
```

Reply:
> Taking logs and then exps is not only slow but it may also lose precision (not in this computation, but still).
> 
> also, more generally, please use braces when using if-else clauses.

Thanks, done as your suggestions. But still remain the grain_size should >= 1.